### PR TITLE
[Feature] Support KLD metric and support evaluation for probabilistic models

### PR DESCRIPTION
**Pull request by LeoXing1996**
# Metric Design
Different from gan metrics, probabilistic metrics:
1. calculated with reconstruction tasks, and can 
2. calculated by a group of parameters but not fake and real images

Therefore, we design a list called `probabilistic_metric_name` to contain probabilistic metrics in `single_gpu_online_evaluation` . When evaluation with those metric separately.

When evaluating those probabilistic metrics, we use set forward mode as **reconstruction**.
We default that all **probabilistic model**  all support this mode. Unlike `forward_test`  which starts with random noise, `mode=reconstruction` performs a reconstruction behavior with given data and returns a dict containing desired probabilistic parameters.

We also slightly modify the batch truncation operation to make the `Metric.feed` support dict input.

## Some further design about `mode=reconstruction`
Although we have not implemented any code, we find a specific function to release reconstruction operation is critical for probabilistic models (e.g., DDPM).
In `train_step`, reconstruction, loss calculation, and update operation are performed altogether.
For `forward_test`, the interface is fixed and called by `sample_from_noise` to perform a random generation process.
Therefore, we need a function to implement a  separate reconstruction process and return all the intermedia probabilistic parameters. And this function can also be called by `train_step`.

### CODE REVIEW

File: mmgen/core/evaluation/metrics.py [hunk 1]
```diff
@@ -383,18 +385,28 @@ def feed(self, batch, mode):
         operation in 'feed_op' function.
 
         Args:
-            batch (Tensor): Images feeded into metric object with order "NCHW"
-                and range [-1, 1].
+            batch (Tensor|dict[Tensor]): Images or dict to be feeded into
```
Comment (nbei) [inline mmgen/core/evaluation/metrics.py:388]: `Tensor | dict`

File: mmgen/core/evaluation/metrics.py [hunk 2]
```diff
@@ -1343,3 +1359,109 @@ def __next__(self):
...
+        if dist.is_initialized():
+            ws = dist.get_world_size()
+            placeholder = [torch.zeros_like(kld) for _ in range(ws)]
+            dist.all_gather(placeholder, kld)
+            kld = torch.cat(placeholder, dim=0)
+
+        # in distributed training, we only collect features at rank-0.
+        if (dist.is_initialized()
+                and dist.get_rank() == 0) or not dist.is_initialized():
+            self.KLD.append(kld.cpu())
```
Comment (nbei) [inline mmgen/core/evaluation/metrics.py:1448]: In general, we won't use capital characters in variables.

**Review by nbei — APPROVED** [review]
